### PR TITLE
gitlab-ci: add CI stage with check shell scripts with shellcheck linter

### LINKED ISSUE #915: [TASK] Perform shellcheck on shell scripts
Shell scripts are risky, but the `shellcheck` tool can help to find issues with it.
- [x] (merged) Delete deprecated scripts
- [x] (**all on review**) Fix existing shellcheck issues
          (number of shellcheck issues)
   - [x] ( 1) ./ci/owncloud/upload_to_owncloud.sh
   - [x] ( 1) ./ci/run_easymesh_cert.sh
   - [x] ( 2) ./docker-builder.sh
   - [x] ( 2) ./docker-runner.sh
   - [x] ( 4) ./framework/platform/bpl/uci/scripts/platform_daemon.sh
   - [x] ( 1) ./tests/openwrt/test_status.sh
   - [x] ( 2) ./tests/test_gw_repeater.sh
   - [x] ( 6) ./tools/docker/builder/openwrt/build.sh
   - [x] ( 1) ./tools/docker/builder/rdkb/toolchain/bitbake.sh
   - [x] ( 6) ./tools/docker/build.sh
   - [x] ( 1) ./tools/docker/clang-format.sh
   - [x] (18) ./tools/docker/image-build.sh
   - [x] (10) ./tools/docker/image-pull.sh
   - [x] ( 1) ./tools/docker/runner/entrypoint.sh
   - [x] ( 3) ./tools/docker/run.sh
   - [x] ( 4) ./tools/docker/static-analysis/cppcheck.sh
   - [x] ( 2) ./tools/docker/stop.sh
   - [x] ( 7) ./tools/docker/test.sh
   - [x] ( 3) ./tools/functions.sh
   - [x] ( 2) ./tools/klocwork/kw.sh

- [x] (Canceled) Add shellcheck to the builder docker image
- [x] (**on review**) Run shellcheck in gitlab CI.

## Total lists
**Note:** Lists below will be updated based on discussion
Removal candidates:
-`./build/common/beerocks/scripts/prplmesh_utils.sh`
-`./build/install/scripts/prplmesh_utils.sh`
-`./build/out/scripts/prplmesh_utils.sh`
-`./common/beerocks/scripts/dhcp.sh`
-`./docker-builder.sh`
-`./docker-runner.sh`
-`./tests/openwrt/test_status.sh`
-`./tools/maptools.sh`
-`./tools/scripts/prplMesh-platform-init.sh`
-`./tools/commands/deploy_rdkb.sh`
-`./tools/commands/deploy_ugw.sh`
-`./framework/platform/bpl/uci/scripts/platform_daemon.sh`

No action due to soon removal:
-`./tests/test_flows.sh`
-`./tests/test_gw_repeater.sh`

Scripts not mentioned in any of lists above will be fixed.

### PR DESCRIPTION
**Pull request by MuteSpirit**
Fix currently present issues detected by shellcheck in prplMesh shell scripts is not enough to keep scripts clean.

GitLab CI should check each commit.

CI will check all shell scripts under git source control.
Calculation diff and detection new issues will be not needed when all PRs with their fixes will be merged: #1123, #1132, #1133, #1204

### CODE REVIEW

**Review by tomereli — APPROVED** [review]

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -56,6 +56,13 @@ flake8-diff:
     - apk add --update --no-cache git
     - git diff -U0 origin/master | flake8 --verbose --diff
 
+shellcheck:
+  stage: build
+  image: pipelinecomponents/shellcheck:0.5.4
+  script:
+    - apk add --update --no-cache git
+    - git ls-files --ignore --exclude '*.sh' | xargs shellcheck -x -P SCRIPTDIR
```
Comment (rmelotte) [inline .gitlab-ci.yml]:
```suggestion
    - git ls-files '*.sh' | xargs shellcheck -x -P SCRIPTDIR
```

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -56,6 +56,13 @@ flake8-diff:
     - apk add --update --no-cache git
     - git diff -U0 origin/master | flake8 --verbose --diff
 
+shellcheck:
+  stage: build
+  image: pipelinecomponents/shellcheck:0.5.4
```
Comment (rmelotte) [inline .gitlab-ci.yml]:
```suggestion
  image: koalaman/shellcheck-alpine:v0.7.1
```

**Review by rmelotte — APPROVED** [review]

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -37,7 +37,7 @@ cppcheck-diff:
     expire_in: 4 weeks
     when: always
   tags:
-    - docker
+- docker
```
Comment (arnout) [inline .gitlab-ci.yml]: That's not good...

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -58,9 +58,10 @@ flake8-diff:
 
 shellcheck:
   stage: build
-  image: pipelinecomponents/shellcheck:0.5.4
-  script:
+  image: koalaman/shellcheck-alpine:v0.7.1
```
Comment (arnout) [inline .gitlab-ci.yml:61]: Can you explain (e.g. in the commit message) why you choose this one?

File: .gitlab-ci.yml (hunk 1 above)
Comment (arnout) [inline .gitlab-ci.yml]: Please use the -z option and xargs -0, to support spaces if we ever need to add those.

The option is called --ignored, not --ignore. And what you're doing there is crazy: you ask to show ignored files and give an additional ignore pattern of *.sh, which indeed gives you all the *.sh files (ignored or not).

The pattern doesn't capture prplmesh_utils.sh.in.

So:
```suggestion
    - git ls-files -z '*.sh' '*.sh.in' | xargs -0 shellcheck -x -P SCRIPTDIR
```

**Review by arnout — APPROVED** [review]

**Comment by arnout** [comment]
The commit message is pretty crappy for various reasons. How about this:

```
 gitlab-ci: add shellcheck job

We use shellcheck to sanitize the shell scripts. Run shellcheck in CI
to make sure any new or modified scripts still comply.

The -P option makes sure that `shellcheck source` paths are relative
to the file itself.

Since some scripts are generated, so include *.sh.in in the shell script
patterns. Theoretically it is possible that such an input file doesn't have
valid shell syntax, but in practice it turns out to be OK.
```

File: .gitlab-ci.yml (hunk 4 above)
Comment (abelog) [reply .gitlab-ci.yml:61]: Different versions of `shellcheck` report different issues for some code. So @MuteSpirit  choose the version  he used.

**Comment by abelog** [comment]
Should be merged after #1204

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -56,6 +56,14 @@ flake8-diff:
     - apk add --update --no-cache git
     - git diff -U0 origin/master | flake8 --verbose --diff
 
+shellcheck:
+  stage: build
+  image: koalaman/shellcheck-alpine:v0.7.1
+  before_script:
+    - apk add --update --no-cache git
+  script:
+    - git ls-files -z '*.sh' '*.sh.in'  | xargs -0 shellcheck -x
```
Comment (arnout) [inline .gitlab-ci.yml]: What happened to the -P option?

Comment (abelog) [reply .gitlab-ci.yml]: I haven't found description of this option in man page of shellcheck version 0.4.3 (which i have on Ubuntu).
Double check using more fresh version (0.7.0) let me found it. So i brought it back.

File: .gitlab-ci.yml (hunk 3 above)
Comment (abelog) [reply .gitlab-ci.yml]: Fixed!

**Comment by abelog** [comment]
*Note*:
Currently, this branch is rebased on top of #1204 . That why diff has more than 1 commit. After #1204 merge diff should be restored to normal state.
Actual change, which requires review is  ea41b1932c1437226ba542f084ec0b427d0e59f9

File: .gitlab-ci.yml (hunk 1 above)
Comment (abelog) [reply .gitlab-ci.yml]: Applied!

File: .gitlab-ci.yml (hunk 2 above)
Comment (abelog) [reply .gitlab-ci.yml]: Applied!

File: .gitlab-ci.yml (hunk 1 above)
Comment (abelog) [reply .gitlab-ci.yml]: Applied!

**Pull request merged**